src/transformers/cache_utils.py

```diff
@@ -186,6 +315,135 @@ def from_legacy_cache(cls, past_key_values: Optional[Tuple[Tuple[torch.FloatTens
+
+        self.key_cache: List[torch.Tensor] = []
+        self.value_cache: List[torch.Tensor] = []
+        self._key_cache_quant: List[torch.Tensor] = []
+        self._value_cache_quant: List[torch.Tensor] = []
+        self.seen_token = 0
+
+        self.residual_length = residual_length
+        self.qtype = qint4 if nbits == 4 else qint2
+        self.q_group_size = q_group_size
+
+    def __getitem__(self, layer_idx: int) -> List[Tuple[torch.Tensor]]:
+        """
+        Support for backwards-compatible `past_key_value` indexing, e.g. `past_key_value[0][0].shape[2]` to get the
+        sequence length.
+        """
+        if layer_idx < len(self):
+            return (self.key_cache[layer_idx], self.value_cache[layer_idx])
+        else:
+            raise KeyError(f"Cache only has {len(self)} layers, attempted to access layer with index {layer_idx}")
+
+    def __iter__(self):
+        """
+        Support for backwards-compatible `past_key_value` iteration, e.g. `for x in past_key_value:` to iterate over
+        keys and values
+        """
+        for layer_idx in range(len(self)):
+            yield (self.key_cache[layer_idx], self.value_cache[layer_idx])
+
+    def __len__(self):
+        """
+        Support for backwards-compatible `past_key_value` length, e.g. `len(past_key_value)`. This value corresponds
+        to the number of layers in the model.
+        """
+        return len(self.key_cache)
+
+    def update(
+        self,
+        key_states: torch.Tensor,
+        value_states: torch.Tensor,
+        layer_idx: int,
+        cache_kwargs: Optional[Dict[str, Any]] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        # Update the number of seen tokens
+        if layer_idx == 0:
+            self.seen_token += key_states.shape[-2]
+
+        if len(self.key_cache) <= layer_idx:
+            self._key_cache_quant.append(self._quantize(key_states.contiguous()))
+            self._value_cache_quant.append(self._quantize(value_states.contiguous()))
+            self.key_cache.append(torch.zeros(0, dtype=key_states.dtype, device=key_states.device))
+            self.value_cache.append(torch.zeros(0, dtype=key_states.dtype, device=key_states.device))
+            keys_to_return, values_to_return = key_states, value_states
+        else:
+            dequant_key = self._key_cache_quant[layer_idx].dequantize()
+            dequant_value = self._value_cache_quant[layer_idx].dequantize()
+            keys_to_return = torch.cat(
+                [
+                    dequant_key,
+                    self.key_cache[layer_idx],
+                    key_states,
+                ],
+                dim=-2,
+            )
+            values_to_return = torch.cat(
+                [
+                    dequant_value,
+                    self.value_cache[layer_idx],
+                    value_states,
+                ],
+                dim=-2,
+            )
```

Comment:
let's split this in a few lines

Reply:
Sorry, I didnt understand what you mean. Format at one line since it's already one variable per line now?

Reply:
```suggestion
            keys_to_return = [dequant_key, self.key_cache[layer_idx],key_states]
            keys_to_return = torch.cat(keys_to_returndim=-2)
```
something like that makes it more readable ! 😉

Reply:
Ahh, okay, thx